src/v/archival/archival_policy.cc

```diff
@@ -10,90 +10,118 @@
 
 #include "archival/archival_policy.h"
 
+#include "archival/logger.h"
 #include "storage/disk_log_impl.h"
+#include "storage/fs_utils.h"
 #include "storage/segment.h"
 #include "storage/segment_set.h"
+#include "storage/version.h"
+#include "vlog.h"
+
+#include <seastar/core/shared_ptr.hh>
 
 namespace archival {
 
-class arch_policy_non_compacted_impl {
-public:
-    arch_policy_non_compacted_impl(
-      model::ntp ntp, const model::revision_id& rev)
-      : _ntp(std::move(ntp))
-      , _rev(rev) {}
+std::ostream& operator<<(std::ostream& s, const upload_candidate& c) {
+    s << "{ exposed_name: " << c.exposed_name
+      << ", starting_offset: " << c.starting_offset
+      << ", segment_file_name: " << c.source->reader().filename() << " }";
+    return s;
+}
 
-    std::optional<manifest> make_local_manifest(storage::log_manager& lm) {
-        manifest tmp(_ntp, _rev);
-        std::optional<storage::log> log = lm.get(_ntp);
-        if (!log) {
-            return std::nullopt;
-        }
-        auto plog = dynamic_cast<storage::disk_log_impl*>(log->get_impl());
-        if (plog == nullptr) {
-            return std::nullopt;
-        }
-        for (const auto& segment : plog->segments()) {
-            bool closed = !segment->has_appender();
-            bool compacted = segment->is_compacted_segment();
-            if (!closed || compacted) {
-                continue;
+archival_policy::archival_policy(model::ntp ntp)
+  : _ntp(std::move(ntp)) {}
+
+upload_candidate archival_policy::get_next_candidate(
+  model::offset last_offset, storage::log_manager& lm) {
+    vlog(
+      archival_log.debug,
+      "Upload policy for {} invoked, last-offset: {}",
+      _ntp,
+      last_offset);
+    std::optional<storage::log> log = lm.get(_ntp);
+    if (!log) {
+        vlog(archival_log.debug, "Upload policy for {} no such ntp", _ntp);
+        return {};
+    }
+    auto plog = dynamic_cast<storage::disk_log_impl*>(log->get_impl());
```

Comment:
add a comment on why we need to access this. 

Also, @dotnwat i wonder if we should start to move away from the in-memory log impl or update it. 
give it a more full interface - as a slow back-burner type of ticket.

Reply:
@senior7515 was thinking about that recently too. I think sometime this year while we are cleaning up some of the storage we can also get the memory log working again. it is really good for test speed / catching bugs similar to building with gcc and clang.